pgjdbc/src/main/java/org/postgresql/util/internal/IntSet.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2024, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.util.internal;
+
+import org.postgresql.core.Oid;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Read-optimized {@code Set} for storing {@link Oid} values.
+ */
+public class IntSet {
+  /**
+   * Maximal Oid that will bs stored in {@link BitSet}.
+   * If Oid exceeds this value, then it will be stored in {@code Set<Int>} only.
+   * In theory, Oids can be up to 32bit, so we want to limit per-connection memory utilization.
+   * Allow {@code BitSet} to consume up to 8KiB (one for send and one for receive).
+   */
+  private static final int MAX_OID_TO_STORE_IN_BITSET = 8192 * 8;
+
+  /**
+   * Contains values outside [0..MAX_OID_TO_STORE_IN_BITSET] range.
+   * This field is null if bitSet contains all the values.
+   */
+  private @Nullable Set<Integer> set;
+
+  /**
+   * Contains values in range of [0..MAX_OID_TO_STORE_IN_BITSET].
+   */
+  private final BitSet bitSet = new BitSet();
+
+  /**
+   * Clears the contents of the set.
+   */
+  public void clear() {
+    set = null;
+    bitSet.clear();
+  }
+
+  /**
+   * Adds all the values to the set.
+   * @param values set of values to add
```

Comment:
Perhaps a comment that null values are not supported.

Reply:
The value is not nullable (`Collection<? extends Integer>` vs `Collection<? extends @Nullable Integer>`)